[DPE-5531] Charmed Spark Hardening guide (#51)
canonical/spark-k8s-bundle: releases/3.4/yaml/docs/explanation/e-hardening.md

| Spark Job (Driver) - GPU Support   | [`charmed-spark-gpu`](https://github.com/orgs/canonical/packages/container/package/charmed-spark-gpu)       |
| Spark Job (Executor) - GPU Support | [`charmed-spark-gpu`](https://github.com/orgs/canonical/packages/container/package/charmed-spark-gpu)       |
| Integration Hub                    | [`spark-integration-hub`](https://github.com/orgs/canonical/packages/container/package/spark-integration-hub)                    |

New versions of the Charmed Spark images may be released to provide patching of vulnerabilities (CVEs). 

### Charm Operator Security Upgrades

Charmed Spark K8s operators, including Spark History server, Kyuubi, and Integration Hub, install a pinned revision of the 
Charmed Spark images outlined in the previous table in order to provide reproducible and secure environments. 
New versions of Charmed Spark K8s operators may therefore be released to provide patching of vulnerabilities (CVEs). 
It is important to refresh the charm regularly to make sure the workload is as secure as possible. 
For more information on how to refresh the charm, see the [how-to upgrade](TODO) guide.

### Encryption

> We recommend to deploy Charmed Spark K8s with encryption enabled for the communication between components whenever available and supported.

Reviewer: Looks like something is missing here. Maybe a verb after components?

In the following, we provide further information on how to encrypt the various data flow between the different components of the solution:

* Client <> Kubernetes API connections

> * Object Storage connections

Reviewer: ```suggestion
* Object storage connections
```

and mutates incoming HTTP(s) requests, fully-integrated with the Canonical Identity Platform.

Refer to the [how-to enable authorization in the Spark history server](/t/charmed-spark-k8s-documentation-how-to-enable-authentication-on-the-spark-history-server-charm/13563) user guide 
for more information. From a permission-wise point of view, white lists of authorised users 
can be provided using the Spark History Server [charm configuration option](https://charmhub.io/spark-history-server-k8s/configurations?channel=3.4/edge). 

#### Kyuubi JDBC endpoint

Authentication can be enabled for Kyuubi via its integration with PostgreSQL charm on the `auth-db` interface. Currently only one admin user is enabled, whose credentials can be retrieved 
using the [`get-jdbc-endpoint` action](https://charmhub.io/kyuubi-k8s/actions#get-jdbc-endpoint).

### Monitoring and Auditing

Charmed Spark provides native integration with the [Canonical Observability Stack (COS)](https://charmhub.io/topics/canonical-observability-stack).
To reduce the blast radius of infrastructure disruptions, the general recommendation is to deploy COS and the observed application into 

> separate environments, isolated one another. Refer to the [COS production deployments best practices](https://charmhub.io/topics/canonical-observability-stack/reference/best-practices)

Reviewer: ```suggestion
separate environments, isolated from one another. Refer to the [COS production deployments best practices](https://charmhub.io/topics/canonical-observability-stack/reference/best-practices) page
```

for more information. 

> Refer to How-To user guide for more information on [how to integrate and customize](/t/charmed-spark-k8s-documentation-enable-monitoring/13063) 

Reviewer: ```suggestion
For more information on how to enable and customise monitoring with COS, see the [Guide](/t/charmed-spark-k8s-documentation-enable-monitoring/13063).
```

> the Charmed Spark deployment with COS. 

Reviewer: ```suggestion
```

> ## Additional Resources

Reviewer: ```suggestion
## Additional resources
```

> For further information and details on the security and cryptographic specifications used by Charmed Spark, please refer to the [Security Explanation page](TODO).

Reviewer: Add the valid link here

Author: Oh, yes. When I initially wrote this, the security docs was not out yet. Updated! Nice catch!
